### PR TITLE
Add support for configuring TLS ServerName for health checks

### LINKED ISSUE #9473: Support configuring ServerName for HTTP health checks
#### Feature Description

Some TLS servers require SNI, but the Golang HTTP client doesn't include it in the ClientHello when [connecting to an IP address](https://github.com/golang/go/blob/4fd94558820100129b98f284e21b19fc27a99926/src/crypto/tls/common.go#L607).

#### Use Case(s)

We ran into this use case while trying to health check external services using consul-esm. Things work fine if we register them using hostnames as addresses, but if we use IP addresses the SNI is not set and handshakes fail.

In #3936 there was a suggestion to add a new `TLSServerName` field to the check definition: https://github.com/hashicorp/consul/issues/3936#issuecomment-442429284. I believe the same `TLSServerName` field could be used here to optionally set `TLSClientConfig.ServerName`, and the change would be backwards compatible. If that sounds reasonable, I'd be happy to provide a PR addressing this.

### PR DESCRIPTION
**Pull request by cbroglie**
Some TLS servers require SNI, but the Golang HTTP client doesn't
include it in the ClientHello when connecting to an IP address. This
change adds a new TLSServerName field to health check definitions to
optionally set it. This fixes #9473.

### CODE REVIEW

**Review by dnephin — COMMENTED** [review]
Thank you for the PR!

I think this change looks good, and the implementation looks both correct and complete. Nice work!

The only concern I have with this change is the refactor in `http_decode_test.go`.  I appreciate the desire to reduce some of the boilerplate here, but I think these tests are already a quite difficult to understand, and with the reduced boilerplate I think unfortunately it makes then even harder to understand.

I suspect another way we can approach these tests is by defining two JSON inputs that set all the fields, one with all CamelCase keys, one with all  snake_case keys, and use those two inputs to test the permutations. There will need to be one other test case for the one field that has more then 2 options, but that's probably ok.

I don't think we need to make that change in this PR, but I think it would be good to use the existing test structure until we can make that change. Alternatively if you wanted to try out that approach first and then introduce this additional field, that would be great as well.

What do you think about that approach?

File: tlsutil/config.go [hunk 1]
```diff
@@ -711,21 +711,27 @@ func (c *Configurator) IncomingHTTPSConfig() *tls.Config {
...
 	c.log("OutgoingTLSConfigForCheck")
+
+	if serverName == "" {
+		serverName = c.serverNameOrNodeName()
+	}
+
 	if !c.enableAgentTLSForChecks() {
 		return &tls.Config{
 			InsecureSkipVerify: skipVerify,
+			ServerName:         serverName,
```
[the remark this answers is not in the document]
Comment (cbroglie) [reply tlsutil/config.go:728]: I did make this change intentionally: the client may not care to verify which cert the server is returning, but the server may still need the SNI to determine which cert to use.

> I'm not sure if this is likely to cause problems or not.

I don't have a good sense of this either. I'd argue my change is good for consistency of SNI handling, but it's not required for my use case and I can revert it if we're worried about the behavior change.

**Comment by cbroglie** [comment]
> I appreciate the desire to reduce some of the boilerplate here, but I think these tests are already a quite difficult to understand, and with the reduced boilerplate I think unfortunately it makes then even harder to understand.

Yep that's fair, I was just optimizing for future authors adding new fields. I'll revert my changes and take a look at either implementing your suggestion or just carrying forward the existing copy/paste.

File: website/content/api-docs/agent/check.mdx [hunk 2]
```diff
@@ -188,6 +188,11 @@ The table below shows this endpoint's support for
   The value can be further limited for all checks of a given agent using the
   `check_output_max_size` flag in the agent.
 
+- `TLSServerName` `(string: "")` - Specifies an optional string used to set the
```
Comment (dnephin) [inline website/content/api-docs/agent/check.mdx:191]: These docs looks good. I think we also need to document this in `website/content/docs/discovery/checks.mdx` , which is linked from the config docs.

**Review by dnephin — COMMENTED** [review]
Sounds good, thank you for working on this!

File: tlsutil/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnephin) [reply tlsutil/config.go:728]: I'll look into this some more to see if I can figure out how most servers handle this field. This change may very well be safe.

**Comment by cbroglie** [comment]
@dnephin I believe I've made all the requested changes and rebased, let me know if anything is missing

**Review by dnephin — APPROVED** [review]
Thank you! This looks good to me.

I noticed that CI was not running on this PR. I believe that is because you have CircleCI enabled on your fork, so the jobs are running as part of your project and for some reason are not being linked back to github.

Once https://app.circleci.com/pipelines/github/cbroglie/consul looks good, I think this is ready to merge.

**Comment by cbroglie** [comment]
Thanks!

File: tlsutil/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hynek) [reply tlsutil/config.go:728]: JFTR, this broke the use-case of using Vault as a CA for Nomad services.

The service health checks used to verify the certificate against the service IP so we put it into `ip_sans`. As of Consul 1.10 it checks against the unqualified hostname so they all got a mismatch and I had to change all my Nomad plans to use `$alt_names := env "attr.unique.consul.name" | printf "alt_names=%s" -}}` and re-deploy.

I'm surprised this didn't break more people's clusters but everybody is probably running with TLS verification off anyways. ;)

[the remark this answers is not in the document]
Comment (dnephin) [reply tlsutil/config.go:728]: I believe #10481 is reporting this same issue, and should be fixed by #10490

[the remark this answers is not in the document]
Comment (beautifulentropy) [reply tlsutil/config.go:728]: Inclusion of the Consul server/ node name in the ServerName field of a client TLS config used exclusively for health checking doesn't appear to be a desirable behavior. I've opened #17481 to fix this and provide an explanation.

**Pull request merged**